### PR TITLE
Remove callback support and make process context async.

### PR DESCRIPTION
**Pull request by dlongley**
This PR removes deprecated callback support with an aim towards supporting `options` in document loaders. It also makes processing contexts asynchronous with an aim towards allowing contexts to load during processing and to enable better active context caching and reuse based on contextual information during context processing.

### CODE REVIEW

File: lib/compact.js [hunk 1]
```diff
@@ -51,9 +51,9 @@ module.exports = api;
  * @param options the compaction options.
  * @param compactionMap the compaction map to use.
  *
- * @return the compacted value.
+ * @return a promise that resolves to the compacted value.
  */
-api.compact = ({
+api.compact = async ({
```
Comment (mattcollier) [inline lib/compact.js]: I assume  this maps directly to `jsonld.compact` in which case this breaking change from sync to async warrants a mention in the changelog.

File: lib/expand.js [hunk 2]
```diff
@@ -63,7 +63,7 @@ module.exports = api;
  *
  * @return a Promise that resolves to the expanded value.
  */
-api.expand = ({
+api.expand = async ({
```
Comment (mattcollier) [inline lib/expand.js]: looks like another important change for the changelog.

File: lib/compact.js (hunk 1 above)
Comment (dlongley) [reply lib/compact.js]: It does not map directly. `jsonld.compact` was already async. The internals were not -- so this is not a public facing change. Same for the other internals.

File: lib/expand.js (hunk 2 above)
Comment (dlongley) [reply lib/expand.js]: See above. These are internal functions that were previously synchronous. The external ones were already asynchronous. That being said, there *is* a separate change that removes callback support on the public API, but the CHANGELOG covers that already.

**Comment by dlongley** [comment]
@mattcollier,

> Does the readme need to be updated before release as well?

Yes, but I don't think we'll do a release immediately after this PR is in. I have a few other changes I'm working on that I want to get into the release that are based off of this PR.

File: tests/graph-container.js [hunk 3]
```diff
@@ -18,8 +18,11 @@ describe('@graph container', () => {
         "value": "x"
       }
     };
-    jsonld.expand(doc, (err, expanded) => {
-      assert.ifError(err);
+    const p = jsonld.expand(doc);
+    assert(p instanceof Promise);
+    p.catch(e => {
```
Comment (davidlehn) [inline tests/graph-container.js]: Could change many/all of these tests to async functions and return a promise.  Avoids needing to call done().

**Comment by dlongley** [comment]
Updated README.

File: tests/graph-container.js (hunk 3 above)
Comment (dlongley) [reply tests/graph-container.js]: @davidlehn, yup, but was just copying the convention in the file. Another PR could do that clean up later.

File: lib/jsonld.js [hunk 4]
```diff
@@ -613,34 +605,10 @@ jsonld.fromRDF = util.callbackify(async function(dataset, options) {
...
-        }
-      });
-    });
-  } else {
-    parsedDataset = Promise.resolve(rdfParser(dataset));
-  }
-
-  parsedDataset = await parsedDataset;
-
-  // back-compat with old parsers that produced legacy dataset format
```
Comment (davidlehn) [inline lib/jsonld.js]: Is this removed on purpose, and if so, should it have a changelog comment?

Comment (dlongley) [reply lib/jsonld.js]: I added a note about it.

File: README.md [hunk 5]
```diff
@@ -181,120 +181,87 @@ const context = {
...
 
 ```js
 // deserialize N-Quads (RDF) to JSON-LD
-jsonld.fromRDF(nquads, {format: 'application/n-quads'}, (err, doc) => {
-  // doc is JSON-LD
-});
-
-// or using promises
-const doc = await jsonld.fromRDF(nquads, {format: 'application/n-quads'});
+const doc = jsonld.fromRDF(nquads, {format: 'application/n-quads'});
```
Comment (davidlehn) [inline README.md]: await?

Comment (dlongley) [reply README.md]: Added

**Comment by dlongley** [comment]
@davidlehn, ok, I'm merging this. If you think we need more information on how to upgrade, please add in another PR.

**Pull request merged**